### ISSUE TITLE
Keep removing old messages from the DOM, until user scrolls up again.

### ISSUE THREAD

**Issue by engelgabriel** [RocketChat/Rocket.Chat#31]
##

**Comment by rodrigok** [comment]
It is still necessary?

**Comment by engelgabriel** [comment]
I think so.. not that important, but nice to have to keep performance as good as possible, specially on mobile, or when you leave the app open for weeks.

**Comment by geekgonecrazy** [comment]
I need to do some more checking.. But I think this might be causing some performance issues on the Electron / Cordova apps.

**Comment by engelgabriel** [comment]
This is becoming more important, see https://github.com/RocketChat/Rocket.Chat/issues/2505

**Comment by engelgabriel** [comment]
This is still necessary

**Comment by lucasvanhalst** [comment]
Yeah it would be great, after a few hundred messages it gets really unresponsive and is quite annoying to use

**Comment by Sing-Li** [comment]
wargame and others on #support are still experiencing and dealing with this daily

**Comment by engelgabriel** [comment]
this is related https://github.com/RocketChat/Rocket.Chat/issues/5111

**Comment by shtefcs** [comment]
Honestly I am surprised that this is a still issue. 

My RC instance freeze when I search for message and **jump to it** .

It past 2 years from the issue report, i think you should really work on that guys.

**Comment by lucasvanhalst** [comment]
This should be top priority if you ask me... It becomes unusable after it has been open for a few hours.

**Comment by Lemmmy** [comment]
I've noticed that this issue is much more prominent than Firefox than Chrome - in Chrome it is almost entirely a non-issue.

Is there any effort to put forward an implementation of this?

Are you looking to keep the scrollbar consistent when loading/unloading messages?

If so, the best way would be to keep a certain amount of messages beyond the viewport loaded, and then to keep the scrollbar consistent, keep track of the height of all messages that have ever been loaded, and put two dummy elements above and below the messages which you update the height on. 

There is a library called [clusterize.js](https://clusterize.js.org/) which uses a technique similar to this, however, it should be noted this specific library does not work with variable height items. However, that doesn't mean it's impossible to implement with a similar method.

**Comment by SeanPackham** [comment]
@rodrigok is this even still relevant with all our new UI updates?

**Comment by sampaiodiego** [comment]
this is still relevant @SeanPackham . new UI elements haven't touched messages yet, but still I don't we'll fix this until we change the whole UI to React.

**Comment by SeanPackham** [comment]
Thanks @sampaiodiego, could we add some more detail to the issue saying it would be fixed when we move to React and mention a milestone?

**Comment by sampaiodiego** [comment]
@karlprieb any plans to start rewriting the UI using React?

**Comment by karlprieb** [comment]
We plan to start the rewriting in the next month, but can take months to finish that.

**Comment by Vringe** [comment]
I think this is still an issue. After updating from `3.12.7` to `3.18.3` it got much worse. As soon as I scroll up about 2 weeks in a channel, reopening that channel takes about 5-10 seconds depending on how many other channels there are around. We have multiple users complaining about that since the update. It does not matter which client you use (Webapp or Electron Client).

**Comment by danel1** [comment]
Looks like we're facing the same issue as @Vringe described. Any Update or workaround on this @engelgabriel @rodrigok @sampaiodiego ? This has a quite critical performance / usability impact for most of our users.

**Comment by hugocostadev** [comment]
After the Room Rewrite to React and Blaze removal in **6.x Release** this is not happening anymore. Closing it

**Comment by computerquip-work** [comment]
Over 10 years now, is this still an issue?

**Comment by ggazzo** [comment]
The problem is that we relied heavily on DOM elements, which virtualization simply removes. We had several issues preventing us from switching, and we had to do several refactorings first. We changed the framework (Blaze -> React), changed how collections work (Minimongo -> Zustand), and removed some imperative procedures (Jump, Select) to use declarative ones. But I'd say that today (at least for the last 4 months) our code would allow virtualization with minimal headaches. I can try something for the next releases.

**Comment by ggazzo** [comment]
closed by https://github.com/RocketChat/Rocket.Chat/pull/40105
